x-pack/plugins/security_solution/server/utils/promises_pool.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+interface PromisePoolArgs<Item, Result> {
+  concurrency?: number;
+  items: Item[];
+  executor: (item: Item) => Promise<Result>;
+}
+
+/**
+ * Runs promises in batches. It ensures that the number of running async tasks
+ * doesn't exceed the concurrency parameter passed to the function.
+ *
+ * @param concurrency - number of tasks run in parallel
+ * @param items - array of items to be passes to async executor
+ * @param executor - an async function to be called with each provided item
+ *
+ * @returns Struct holding results or errors of async tasks
+ */
+export const initPromisePool = async <Item, Result>({
+  concurrency = 1,
+  items,
+  executor,
+}: PromisePoolArgs<Item, Result>) => {
```

Comment:
First off, thank you for introducing this utility, it's very cool!
The implementation looks correct at first glance and should be _really_ useful for us 👍 
But could you please cover it with unit tests?

Reply:
Added some tests 👍